File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private: 
    int dfs(TreeNode* root) {
        
        if(!root)
            return 0; 
        
        int left = dfs(root -> left) + 1;
        int right = dfs(root -> right) + 1;
        
        return max(left, right); 
    }
public:
    int diameterOfBinaryTree(TreeNode* root) {
        
        if(!root)
            return 0; 
        
        int op1 = dfs(root -> left) + dfs(root -> right);
        int op2 = diameterOfBinaryTree(root -> left);
        int op3 = diameterOfBinaryTree(root -> right);
        
        
        return std::max(op1, std::max(op2, op3)); 
        
    }
};

File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, root: TreeNode) -> int:
        if(not root):
            return 0
        left = self.dfs(root.left) + 1
        right = self.dfs(root.right) + 1
        
        return max(left, right)
        
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if(not root):
            return 0
        
        op1 = self.dfs(root.left) + self.dfs(root.right)
        op2 = self.diameterOfBinaryTree(root.left)
        op3 = self.diameterOfBinaryTree(root.right)
        
        
        return max(op1, max(op2,op3))